Add GCP function bindings

pericles-tpt opened the pull request:
- implement `GDALGetGCPSpatialRef`, `GDALGetGCPCount`, `GDALGetGCPs`, `GDALGetGCPProjection`, `GDALSetGCPs` and `GDALSetGCPs2` bindings
- implement `GetGCPSpatialRef`, `GetGCPCount`, `GetGCPs`, `GetGCPProjection`, `SetGCPs` and `SetGCPs2` go functions
- add `TestGCPs...` tests to test the bindings

pericles-tpt commented:
We were thinking it may be worthwhile to combine the `godalGetGCPCount` and `godalGetGCPs` bindings into a single binding, to minimise `cgo` calls. We could do this by implementing a cpp function that calls both `GDALGetGCPCount` and `GDALGetGCPs`, then we could return both properties in a custom struct like this:
```
typedef struct {
    GDAL_GCP *gcpList;
    int numGCPs;
} GCPsAndCount;
```


One case where we retrieve both properties in our current Go code is the `GetGCPs()` function which requires them to convert the incoming `*C.GDAL_GCP` to an `[]GCP`

Would you like us to combine those bindings? If so, would you be happy with the approach above, or would you prefer a different one?

tbonfort commented on godal.go:
```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+}
+
+// SetGCPs runs the GDALSetGCPs function
+func (ds *Dataset) SetGCPs(GCPList []GCP, pszGCPProjection string, opts ...SetGCPsOption) error {
+	setGCPsOpts := setGCPsOpts{}
+	for _, opt := range opts {
+		opt.setSetGCPsOpt(&setGCPsOpts)
+	}
+	cgc := createCGOContext(nil, setGCPsOpts.errorHandler)
+
+	GCPProj := C.CString(pszGCPProjection)
+	defer C.free(unsafe.Pointer(GCPProj))
+
+	C.godalSetGCPs(cgc.cPointer(), ds.handle(), C.int(len(GCPList)), goGCPArrayToGDALGCP(GCPList), GCPProj)
+	if err := cgc.close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetGCPs2 runs the GDALSetGCPs2 function
+func (ds *Dataset) SetGCPs2(GCPList []GCP, sr *SpatialRef, opts ...SetGCPsOption) error {
+	setGCPsOpts := setGCPsOpts{}
+	for _, opt := range opts {
+		opt.setSetGCPsOpt(&setGCPsOpts)
+	}
+	cgc := createCGOContext(nil, setGCPsOpts.errorHandler)
+
+	srHandle := C.OGRSpatialReferenceH(nil)
+	if sr != nil {
+		srHandle = sr.handle
+	}
+
+	C.godalSetGCPs2(cgc.cPointer(), ds.handle(), C.int(len(GCPList)), goGCPArrayToGDALGCP(GCPList), srHandle)
+	if err := cgc.close(); err != nil {
+		return err
+	}
+	return nil
+}
```
Given that gdal's setgcp call accepts an empty projectionstring or projectionref, I would suggest that you use only a single golang SetGCPs function
```go func (ds *Dataset) SetGCPs(GCPList []GCP,  opts ...SetGCPsOption) error```
and allow passing the projection string or ref through SetGCPOptions.
i.e. as a user you can call either:

- `ds.SetGCPs(mygcplist)`
- `ds.SetGCPs(mygcplist, godal.GCPProjection("+proj=lonlat +foo=bar"))`
- `ds.SetGCPs(mygcplist, godal.GCPSpatialRef(sr) //sr is a *godal.SpatialRef`

tbonfort commented on godal.go:
```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+}
+
+// GetGCPCount runs the GDALGetGCPCount function
+func (ds *Dataset) GetGCPCount() int {
```
```suggestion
func (ds *Dataset) GCPCount() int {
```

tbonfort commented on godal.go:
```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+
+// GetGCPs runs the GDALGetGCPs function
+// TODO: This makes 2 cgo calls, we could reduce it to 1 by combining `godalGetGCPs` and `GetGCPCount`
+func (ds *Dataset) GetGCPs() []GCP {
```
```suggestion
func (ds *Dataset) GCPs() []GCP {
```

tbonfort commented on godal.go:
```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+}
+
+// GetGCPProjection runs the GDALGetGCPProjection function
+func (ds *Dataset) GetGCPProjection() string {
```
```suggestion
func (ds *Dataset) GCPProjection() string {
```

tbonfort commented on godal.go:
```diff
@@ -3923,6 +3923,133 @@ func (ds *Dataset) NearblackInto(sourceDs *Dataset, switches []string, opts ...N
+}
+
+// GetGCPSpatialRef runs the GDALGetGCPSpatialRef function
+func (ds *Dataset) GetGCPSpatialRef() *SpatialRef {
```
```suggestion
func (ds *Dataset) GCPSpatialRef() *SpatialRef {
```

tbonfort commented on godal.cpp:
```diff
@@ -1766,5 +1766,71 @@ GDALDatasetH godalNearblack(cctx *ctx, const char *pszDest, GDALDatasetH hDstDS,
+
+    for (int i = 0; i < numGCPs; i++)
+    {
+		ret[i].pszId = CPLStrdup(pszIds[i]); 
+		ret[i].pszInfo = CPLStrdup(pszInfos[i]);
```
I believe that the corresponding CPLFrees are never called here?

pericles-tpt replied:
The CPLFrees are called in `godalSetGCPs` and `godalSetGCPs2`, when they call `GDALDeinitGCPs`. I agree with your later comment about calling `GDALInitGCPs` and `GDALDeInitGCPs` in the same function. I will work on reworking `godalSetGCPs` shortly, which should address this.

tbonfort commented on options.go:
```diff
@@ -1254,6 +1254,15 @@ type NearblackOption interface {
 }
 
+type setGCPsOpts struct {
+	errorHandler ErrorHandler
```
```suggestion
	errorHandler ErrorHandler
	projString string
	sr *Spatialref
```
c.f. SetGCPs and SetGCPs2 unification comment

tbonfort commented:
> We were thinking it may be worthwhile to combine the `godalGetGCPCount` and `godalGetGCPs` bindings into a single binding, to minimise `cgo` calls. We could do this by implementing a cpp function that calls both `GDALGetGCPCount` and `GDALGetGCPs`, then we could return both properties in a custom struct like this:
> 
> ```
> typedef struct {
>     GDAL_GCP *gcpList;
>     int numGCPs;
> } GCPsAndCount;
> ```
> 
> One case where we retrieve both properties in our current Go code is the `GetGCPs()` function which requires them to convert the incoming `*C.GDAL_GCP` to an `[]GCP`
> 
> Would you like us to combine those bindings? If so, would you be happy with the approach above, or would you prefer a different one?

- +1 for making changes to the code so that GetGCPs results in a single cgo call
- I would argue that there is no use in exposing GetGCPCount as that value can be obtained from len(GetGCPs). Is there a use-cases where the user does not wish to go through the gcp allocations and just get the count?

tbonfort commented on godal.cpp:
```diff
@@ -1766,5 +1766,71 @@ GDALDatasetH godalNearblack(cctx *ctx, const char *pszDest, GDALDatasetH hDstDS,
+	return;
+}
+
+// Utility function to aid conversion from Go `GCP` type to `GDAL_GCP*`, allocates memory for `GDAL_GCP*` and requires a manual `CPLFree`
```
Ideally, this function should be removed and merged into `godalSetGCPs`, with `godalSetGCPs` signature changed to accept our own C.goGCPList struct instead of a `*GDAL_GCP`. This would also allow calling `GDALInitGCPs` and `GDALDeInitGCPs` in the same function which is cleaner / easier to understand

pericles-tpt commented:
We have now addressed all of the changes and suggestions you mentioned:
- We've merged `SetGCPs2` functionality into `SetGCPs` and have added handling in `SetGCPs` of a passed "projection string" or "ref" through `SetGCPOptions`
- The "Get" prefix has been removed from all of the GCP functions that you've mentioned
- The "CPLFrees" issue is handled in a later fix
- The additional options have been added to `setGCPOpts`, for `projString` and `sr`

- A new `GCPsAndCount` struct was added to allow us to call `GDALGetGCPs()` and `GDALGetGCPCount()` in a single cgo call
- The `GetGCPCount()` go function has been removed, as you mentioned it can be obtained from `len(GetGCPs())`

- The utility function `godalGCPPropertiesToGDALGCP()` and `goGCPArrayToGDALGCP()` have been removed
- A `goGCPList` struct was added, and the `godalSetGCPs` and `godalSetGCPs2` functions now take it as a parameter, so all `GDAL_GCP` allocations and free's are now done inside `godalSetGCPs`/`godalSetGCPs2`
- `SetGCPs` was also updated to construct a `C.goGCPList` from the input `[]GCP`

tbonfort commented on godal.cpp:
```diff
@@ -1767,4 +1767,77 @@ GDALDatasetH godalNearblack(cctx *ctx, const char *pszDest, GDALDatasetH hDstDS,
+	godalWrap(ctx);
+
+	// Convert `goGCPList` -> `GDAL_GCP*`
+	GDAL_GCP *GDALGCPList = static_cast<GDAL_GCP *>(CPLCalloc(numGCPs, sizeof(GDAL_GCP)));
+    GDALInitGCPs(numGCPs, GDALGCPList);
+    for (int i = 0; i < numGCPs; i++)
+    {
+		GDALGCPList[i].pszId = CPLStrdup(GCPList.pszIds[i]); 
+		GDALGCPList[i].pszInfo = CPLStrdup(GCPList.pszInfos[i]);
+        GDALGCPList[i].dfGCPPixel = GCPList.dfGCPPixels[i];
+		GDALGCPList[i].dfGCPLine = GCPList.dfGCPLines[i];
+		GDALGCPList[i].dfGCPX = GCPList.dfGCPXs[i];
+		GDALGCPList[i].dfGCPY = GCPList.dfGCPYs[i];
+		GDALGCPList[i].dfGCPZ = GCPList.dfGCPZs[i];
+    }
```
nitpick 1: this code can be mutualized with godalSetGCPs2, right?
nitpick 2: indentation needs fixing

pericles-tpt replied:
We created a new function to address the duplicate code and fixed the indentation issue here

tbonfort commented on godal.h:
```diff
@@ -154,6 +154,25 @@ extern "C" {
 	void godalGridCreate(cctx *ctx, char *pszAlgorithm, GDALGridAlgorithm eAlgorithm, GUInt32 nPoints, const double *padfX, const double *padfY, const double *padfZ, double dfXMin, double dfXMax, double dfYMin, double dfYMax, GUInt32 nXSize, GUInt32 nYSize, GDALDataType eType, void *pData);
 	GDALDatasetH godalGrid(cctx *ctx, const char *pszDest, GDALDatasetH hSrcDS, char **switches);
 	GDALDatasetH godalNearblack(cctx *ctx, const char *pszDest, GDALDatasetH hDstDS, GDALDatasetH hSrcDS, char **switches);
+
```
nitpick: indentation

pericles-tpt replied:
We're not sure what you mean by "indentation" here, these signatures seem to be in line with the other function signatures. Did you meant that `godalGridCreate()` is too long and needs to be split / indented across multiple lines like `godalDatasetRasterIO()`?

tbonfort replied:
sorry, my bad. In the diff view I thought your additions were misaligned.

tbonfort reviewed (commented):
LGTM with a few cosmetic changes

tbonfort commented:
@pericles-tpt ready to merge ?

pericles-tpt commented:
> @pericles-tpt ready to merge ?

Yes, it's ready

andyzimmerman1 commented:
@tbonfort @pericles-tpt 
I would like to use the new GCP function bindings exposed in this pull request. Do you know when there will be a tagged version that includes these changes?

tbonfort commented:
@andyzimmerman1  v0.0.9 tagged

andyzimmerman1 commented:
@tbonfort Fantastic - thank you so much!!
